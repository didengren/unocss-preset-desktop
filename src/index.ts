import type { Preset } from 'unocss'

export interface PresetDesktopOption {
  /**
   * default number of nodes covered by enter animation
   * @default 6
   */
  maxOutput?: number
}

const defaultOption: Required<PresetDesktopOption> = {
  maxOutput: 6,
}

export function presetDesktop(option?: PresetDesktopOption): Preset {
  const config = Object.assign(defaultOption, option)
  return {
    name: '@trina/unocss-preset-desktop',
    rules: [
      [
        'font-desktop',
        {
          'font-family':
            '-apple-system, BlinkMacSystemFont, \'Segoe UI\', Roboto, \'Helvetica Neue\', Arial, \'SourceHanSans-VF\', \'Noto Sans\', sans-serif, \'Apple Color Emoji\', \'Segoe UI Emoji\', \'Segoe UI Symbol\', \'Noto Color Emoji\'',
        },
      ],
    ],
    theme: {
      breakpoints: {
        'sm': '640px',
        'md': '768px',
        'lg': '1024px',
        'xl': '1280px',
        '2xl': '1536px',
      },
    },
    preflights: [
      {
        getCSS: () => {
          const { maxOutput } = config
          const createCss = (index: number, d = 'x') => {
            const upd = d.toUpperCase()
            return `\n
              *> .enter-${d}:nth-child(${index}){
                transform: translate${upd}(50px);
              }
              *> .-enter-${d}:nth-child(${index}){
                transform: translate${upd}(-50px);
              }
              * > .enter-${d}:nth-child(${index}),* > .-enter-${d}:nth-child(${index}){
                z-index: ${10 - index};
                opacity: 0;
                animation: enter-${d}-animation 0.4s ease-in-out 0.3s;
                animation-fill-mode: forwards;
                animation-delay: ${(index * 1) / 10}s;
              }
            `
          }
          let addRawCss = ''
          for (let index = 1; index < maxOutput; index++) {
            addRawCss += createCss(index, 'x')
            addRawCss += createCss(index, 'y')
          }
          return `
            ${addRawCss}
            @keyframes enter-x-animation {
              to {
                opacity: 1;
                transform: translateX(0);
              }
            }
            @keyframes enter-y-animation {
              to {
                opacity: 1;
                transform: translateY(0);
              }
            }
          `
        },
      },
    ],
  }
}
