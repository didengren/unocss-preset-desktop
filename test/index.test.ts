import { createGenerator, presetAttributify, presetUno } from 'unocss'
import { describe, expect, it } from 'vitest'
import { presetDesktop } from '../src'

describe('preset desktop', () => {
  const generator = createGenerator({
    presets: [
      presetUno(),
      presetAttributify(),
      presetDesktop(),
    ],
  })

  it('font family', async () => {
    const { css } = await generator.generate(['font-desktop'].join(' '))

    expect(css).toMatchSnapshot()
  })
})
