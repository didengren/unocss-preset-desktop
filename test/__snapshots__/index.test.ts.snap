// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`preset desktop > font family 1`] = `
"/* layer: preflights */
*,::before,::after{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x: ;--un-pan-y: ;--un-pinch-zoom: ;--un-scroll-snap-strictness:proximity;--un-ordinal: ;--un-slashed-zero: ;--un-numeric-figure: ;--un-numeric-spacing: ;--un-numeric-fraction: ;--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 rgba(0,0,0,0);--un-ring-shadow:0 0 rgba(0,0,0,0);--un-shadow-inset: ;--un-shadow:0 0 rgba(0,0,0,0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgba(147,197,253,0.5);--un-blur: ;--un-brightness: ;--un-contrast: ;--un-drop-shadow: ;--un-grayscale: ;--un-hue-rotate: ;--un-invert: ;--un-saturate: ;--un-sepia: ;--un-backdrop-blur: ;--un-backdrop-brightness: ;--un-backdrop-contrast: ;--un-backdrop-grayscale: ;--un-backdrop-hue-rotate: ;--un-backdrop-invert: ;--un-backdrop-opacity: ;--un-backdrop-saturate: ;--un-backdrop-sepia: ;}::backdrop{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x: ;--un-pan-y: ;--un-pinch-zoom: ;--un-scroll-snap-strictness:proximity;--un-ordinal: ;--un-slashed-zero: ;--un-numeric-figure: ;--un-numeric-spacing: ;--un-numeric-fraction: ;--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 rgba(0,0,0,0);--un-ring-shadow:0 0 rgba(0,0,0,0);--un-shadow-inset: ;--un-shadow:0 0 rgba(0,0,0,0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgba(147,197,253,0.5);--un-blur: ;--un-brightness: ;--un-contrast: ;--un-drop-shadow: ;--un-grayscale: ;--un-hue-rotate: ;--un-invert: ;--un-saturate: ;--un-sepia: ;--un-backdrop-blur: ;--un-backdrop-brightness: ;--un-backdrop-contrast: ;--un-backdrop-grayscale: ;--un-backdrop-hue-rotate: ;--un-backdrop-invert: ;--un-backdrop-opacity: ;--un-backdrop-saturate: ;--un-backdrop-sepia: ;}

            ul {
              list-style: none;
              margin: 0;
              padding: 0;
            }
            

              *> .enter-x:nth-child(1){
                transform: translateX(50px);
              }
              *> .-enter-x:nth-child(1){
                transform: translateX(-50px);
              }
              * > .enter-x:nth-child(1),* > .-enter-x:nth-child(1){
                z-index: 9;
                opacity: 0;
                animation: enter-x-animation 0.4s ease-in-out 0.3s;
                animation-fill-mode: forwards;
                animation-delay: 0.1s;
              }
            

              *> .enter-y:nth-child(1){
                transform: translateY(50px);
              }
              *> .-enter-y:nth-child(1){
                transform: translateY(-50px);
              }
              * > .enter-y:nth-child(1),* > .-enter-y:nth-child(1){
                z-index: 9;
                opacity: 0;
                animation: enter-y-animation 0.4s ease-in-out 0.3s;
                animation-fill-mode: forwards;
                animation-delay: 0.1s;
              }
            

              *> .enter-x:nth-child(2){
                transform: translateX(50px);
              }
              *> .-enter-x:nth-child(2){
                transform: translateX(-50px);
              }
              * > .enter-x:nth-child(2),* > .-enter-x:nth-child(2){
                z-index: 8;
                opacity: 0;
                animation: enter-x-animation 0.4s ease-in-out 0.3s;
                animation-fill-mode: forwards;
                animation-delay: 0.2s;
              }
            

              *> .enter-y:nth-child(2){
                transform: translateY(50px);
              }
              *> .-enter-y:nth-child(2){
                transform: translateY(-50px);
              }
              * > .enter-y:nth-child(2),* > .-enter-y:nth-child(2){
                z-index: 8;
                opacity: 0;
                animation: enter-y-animation 0.4s ease-in-out 0.3s;
                animation-fill-mode: forwards;
                animation-delay: 0.2s;
              }
            

              *> .enter-x:nth-child(3){
                transform: translateX(50px);
              }
              *> .-enter-x:nth-child(3){
                transform: translateX(-50px);
              }
              * > .enter-x:nth-child(3),* > .-enter-x:nth-child(3){
                z-index: 7;
                opacity: 0;
                animation: enter-x-animation 0.4s ease-in-out 0.3s;
                animation-fill-mode: forwards;
                animation-delay: 0.3s;
              }
            

              *> .enter-y:nth-child(3){
                transform: translateY(50px);
              }
              *> .-enter-y:nth-child(3){
                transform: translateY(-50px);
              }
              * > .enter-y:nth-child(3),* > .-enter-y:nth-child(3){
                z-index: 7;
                opacity: 0;
                animation: enter-y-animation 0.4s ease-in-out 0.3s;
                animation-fill-mode: forwards;
                animation-delay: 0.3s;
              }
            

              *> .enter-x:nth-child(4){
                transform: translateX(50px);
              }
              *> .-enter-x:nth-child(4){
                transform: translateX(-50px);
              }
              * > .enter-x:nth-child(4),* > .-enter-x:nth-child(4){
                z-index: 6;
                opacity: 0;
                animation: enter-x-animation 0.4s ease-in-out 0.3s;
                animation-fill-mode: forwards;
                animation-delay: 0.4s;
              }
            

              *> .enter-y:nth-child(4){
                transform: translateY(50px);
              }
              *> .-enter-y:nth-child(4){
                transform: translateY(-50px);
              }
              * > .enter-y:nth-child(4),* > .-enter-y:nth-child(4){
                z-index: 6;
                opacity: 0;
                animation: enter-y-animation 0.4s ease-in-out 0.3s;
                animation-fill-mode: forwards;
                animation-delay: 0.4s;
              }
            

              *> .enter-x:nth-child(5){
                transform: translateX(50px);
              }
              *> .-enter-x:nth-child(5){
                transform: translateX(-50px);
              }
              * > .enter-x:nth-child(5),* > .-enter-x:nth-child(5){
                z-index: 5;
                opacity: 0;
                animation: enter-x-animation 0.4s ease-in-out 0.3s;
                animation-fill-mode: forwards;
                animation-delay: 0.5s;
              }
            

              *> .enter-y:nth-child(5){
                transform: translateY(50px);
              }
              *> .-enter-y:nth-child(5){
                transform: translateY(-50px);
              }
              * > .enter-y:nth-child(5),* > .-enter-y:nth-child(5){
                z-index: 5;
                opacity: 0;
                animation: enter-y-animation 0.4s ease-in-out 0.3s;
                animation-fill-mode: forwards;
                animation-delay: 0.5s;
              }
            
            @keyframes enter-x-animation {
              to {
                opacity: 1;
                transform: translateX(0);
              }
            }
            @keyframes enter-y-animation {
              to {
                opacity: 1;
                transform: translateY(0);
              }
            }
          
/* layer: default */
.font-desktop{font-family:-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,SourceHanSans-VF,Noto Sans,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol,Noto Color Emoji;}"
`;
